In [2]:
import pandas as pd
import numpy as np

# Importation et nettoyage des données

## Customers

In [3]:
# importation du dataframe customers
customer= pd.read_csv("customers.csv", sep=",")
customer

client_id sex  birth
0       c_4410   f   1967
1       c_7839   f   1975
2       c_1699   f   1984
3       c_5961   f   1962
4       c_5320   m   1943
...        ...  ..    ...
8618    c_7920   m   1956
8619    c_7403   f   1970
8620    c_5119   m   1974
8621    c_5643   f   1968
8622      c_84   f   1982

[8623 rows x 3 columns]

In [4]:
# affichage des clients test
custest=customer[~customer["client_id"].str.contains("c_", na=False)]
print(custest.shape)
custest

(2, 3)


client_id sex  birth
2735      ct_0   f   2001
8494      ct_1   m   2001

#### 2 clients sont des clients tests, on les suppriment donc de la base de données

In [5]:
# dataframe sans les clients tests
customer=customer[customer["client_id"].str.contains("c_", na=False)]
customer.head()

client_id sex  birth
0    c_4410   f   1967
1    c_7839   f   1975
2    c_1699   f   1984
3    c_5961   f   1962
4    c_5320   m   1943

In [6]:
# vérification des types de colonnes et des valeurs nulles
customer.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8621 entries, 0 to 8622
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   client_id  8621 non-null   object
 1   sex        8621 non-null   object
 2   birth      8621 non-null   int64 
dtypes: int64(1), object(2)
memory usage: 269.4+ KB


In [7]:
# vérification des valeurs nulles
customer.isnull().sum()

client_id    0
sex          0
birth        0
dtype: int64

#### Il n'y a plus de valeurs nulles dans le fichier

In [8]:
# création de la colonne âge
customer["age"]= 2021-customer["birth"]

<ipython-input-8-d552edb9dbd2>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  customer["age"]= 2021-customer["birth"]


In [9]:
# nouveau dataframe
customer

client_id sex  birth  age
0       c_4410   f   1967   54
1       c_7839   f   1975   46
2       c_1699   f   1984   37
3       c_5961   f   1962   59
4       c_5320   m   1943   78
...        ...  ..    ...  ...
8618    c_7920   m   1956   65
8619    c_7403   f   1970   51
8620    c_5119   m   1974   47
8621    c_5643   f   1968   53
8622      c_84   f   1982   39

[8621 rows x 4 columns]

In [10]:
# description du dataframe
customer.describe()

birth          age
count  8621.000000  8621.000000
mean   1978.275606    42.724394
std      16.917958    16.917958
min    1929.000000    17.000000
25%    1966.000000    29.000000
50%    1979.000000    42.000000
75%    1992.000000    55.000000
max    2004.000000    92.000000

#### La moyenne d'âge des clients est de 42.72 ans avec des âges allant de 17 à 92 ans

In [11]:
# vérification des valeurs doubles
customer.duplicated().sum()

0

In [12]:
# vérification des clients id doubles
customer["client_id"].duplicated().sum()

0

In [13]:
# verification de la colonne sexe
customer["sex"].unique()

array(['f', 'm'], dtype=object)

In [14]:
 # verification de la clé primaire
customer["client_id"].nunique()==customer.shape[0]

True

#### Client_id est la clé primaire de ce dataframe

## Products

In [15]:
# importation du dataframe products
product= pd.read_csv("products.csv", sep=",")
product

id_prod   price  categ
0     0_1421   19.99      0
1     0_1368    5.13      0
2      0_731   17.99      0
3      1_587    4.99      1
4     0_1507    3.99      0
...      ...     ...    ...
3282    2_23  115.99      2
3283   0_146   17.14      0
3284   0_802   11.22      0
3285   1_140   38.56      1
3286  0_1920   25.16      0

[3287 rows x 3 columns]

In [16]:
# vérification des types de colonnes et des valeurs nulles
product.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3287 entries, 0 to 3286
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id_prod  3287 non-null   object 
 1   price    3287 non-null   float64
 2   categ    3287 non-null   int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 77.2+ KB


In [17]:
product.isnull().sum()

id_prod    0
price      0
categ      0
dtype: int64

#### Il n'y a plus de valeurs nulles dans le fichier

In [18]:
# vérification des valeurs doubles
product.duplicated().sum()

0

In [19]:
# vérification des id_prod doubles
product["id_prod"].duplicated().sum()

0

In [20]:
# analyse des prix
product[["price"]].describe()

price
count  3287.000000
mean     21.856641
std      29.847908
min      -1.000000
25%       6.990000
50%      13.060000
75%      22.990000
max     300.000000

#### Les prix vont de -1 à 300 euros avec une moyenne de 21.85 euros. On voit qu'il y a une erreur sur le prix avec -1

In [21]:
# affichage des produits avec un prix inférieur à 0
product[product["price"]<=0]

id_prod  price  categ
731     T_0   -1.0      0

#### Suppression du prix -1 qui était un produit test

In [23]:
# création du dataframe avec les produits supérieurs à 0
product= product[product["price"]>=0]
product

id_prod   price  categ
0     0_1421   19.99      0
1     0_1368    5.13      0
2      0_731   17.99      0
3      1_587    4.99      1
4     0_1507    3.99      0
...      ...     ...    ...
3282    2_23  115.99      2
3283   0_146   17.14      0
3284   0_802   11.22      0
3285   1_140   38.56      1
3286  0_1920   25.16      0

[3286 rows x 3 columns]

In [24]:
# description des prix du nouveau dataframe
product[["price"]].describe()

price
count  3286.000000
mean     21.863597
std      29.849786
min       0.620000
25%       6.990000
50%      13.075000
75%      22.990000
max     300.000000

In [22]:
# verification de la clé primaire
product["id_prod"].nunique()==product.shape[0]

True

#### id_prod est la clé primaire de ce dataframe

## Transactions

In [23]:
# importation du dataframe transactions
transaction= pd.read_csv("transactions.csv", sep=",")
transaction

id_prod                        date session_id client_id
0       0_1518  2022-05-20 13:21:29.043970   s_211425     c_103
1        1_251  2022-02-02 07:55:19.149409   s_158752    c_8534
2       0_1277  2022-06-18 15:44:33.155329   s_225667    c_6714
3        2_209  2021-06-24 04:19:29.835891    s_52962    c_6941
4       0_1509  2023-01-11 08:22:08.194479   s_325227    c_4232
...        ...                         ...        ...       ...
679527  0_1551  2022-01-15 13:05:06.246925   s_150195    c_8489
679528   1_639  2022-03-19 16:03:23.429229   s_181434    c_4370
679529  0_1425  2022-12-20 04:33:37.584749   s_314704     c_304
679530  0_1994  2021-07-16 20:36:35.350579    s_63204    c_2227
679531   1_523  2022-09-28 01:12:01.973763   s_274568    c_3873

[679532 rows x 4 columns]

In [24]:
# vérification des types de colonnes et des valeurs nulles
transaction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 679532 entries, 0 to 679531
Data columns (total 4 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   id_prod     679532 non-null  object
 1   date        679532 non-null  object
 2   session_id  679532 non-null  object
 3   client_id   679532 non-null  object
dtypes: object(4)
memory usage: 20.7+ MB


In [25]:
# affichage des valeurs nulles
transaction.isnull().sum()

id_prod       0
date          0
session_id    0
client_id     0
dtype: int64

#### Il n'y a plus de valeurs nulles

In [26]:
# vérification des valeurs doubles
transaction.duplicated().sum()

126

#### Il y a 126 valeurs doubles que l'on va observer

In [27]:
# affichage des valeurs doubles
transdoub=transaction[transaction.duplicated()]
transdoub.head()

id_prod                             date session_id client_id
27778      T_0  test_2021-03-01 02:30:02.237437        s_0      ct_1
52424      T_0  test_2021-03-01 02:30:02.237419        s_0      ct_0
96687      T_0  test_2021-03-01 02:30:02.237412        s_0      ct_1
130188     T_0  test_2021-03-01 02:30:02.237419        s_0      ct_0
139339     T_0  test_2021-03-01 02:30:02.237443        s_0      ct_1

In [28]:
# affichage des lignes de test
test_t1=transdoub[transdoub["date"].str.startswith("test")]
test_t1

id_prod                             date session_id client_id
27778      T_0  test_2021-03-01 02:30:02.237437        s_0      ct_1
52424      T_0  test_2021-03-01 02:30:02.237419        s_0      ct_0
96687      T_0  test_2021-03-01 02:30:02.237412        s_0      ct_1
130188     T_0  test_2021-03-01 02:30:02.237419        s_0      ct_0
139339     T_0  test_2021-03-01 02:30:02.237443        s_0      ct_1
...        ...                              ...        ...       ...
653098     T_0  test_2021-03-01 02:30:02.237432        s_0      ct_0
657830     T_0  test_2021-03-01 02:30:02.237417        s_0      ct_0
662081     T_0  test_2021-03-01 02:30:02.237427        s_0      ct_1
671647     T_0  test_2021-03-01 02:30:02.237424        s_0      ct_1
679180     T_0  test_2021-03-01 02:30:02.237425        s_0      ct_1

[126 rows x 4 columns]

#### Les lignes qui étaient doublées étaient des données test, on les suppriment donc

In [29]:
# nouveau dataframe sans les tests
transaction = transaction[~transaction["date"].str.startswith("test")]
print(transaction.shape)
transaction

(679332, 4)


id_prod                        date session_id client_id
0       0_1518  2022-05-20 13:21:29.043970   s_211425     c_103
1        1_251  2022-02-02 07:55:19.149409   s_158752    c_8534
2       0_1277  2022-06-18 15:44:33.155329   s_225667    c_6714
3        2_209  2021-06-24 04:19:29.835891    s_52962    c_6941
4       0_1509  2023-01-11 08:22:08.194479   s_325227    c_4232
...        ...                         ...        ...       ...
679527  0_1551  2022-01-15 13:05:06.246925   s_150195    c_8489
679528   1_639  2022-03-19 16:03:23.429229   s_181434    c_4370
679529  0_1425  2022-12-20 04:33:37.584749   s_314704     c_304
679530  0_1994  2021-07-16 20:36:35.350579    s_63204    c_2227
679531   1_523  2022-09-28 01:12:01.973763   s_274568    c_3873

[679332 rows x 4 columns]

In [30]:
# séparation de l'heure et la date
transaction[["date","heure"]] = transaction.date.str.split(expand=True)
transaction

C:\Users\seb\anaconda3\lib\site-packages\pandas\core\frame.py:3191: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]


id_prod        date session_id client_id            heure
0       0_1518  2022-05-20   s_211425     c_103  13:21:29.043970
1        1_251  2022-02-02   s_158752    c_8534  07:55:19.149409
2       0_1277  2022-06-18   s_225667    c_6714  15:44:33.155329
3        2_209  2021-06-24    s_52962    c_6941  04:19:29.835891
4       0_1509  2023-01-11   s_325227    c_4232  08:22:08.194479
...        ...         ...        ...       ...              ...
679527  0_1551  2022-01-15   s_150195    c_8489  13:05:06.246925
679528   1_639  2022-03-19   s_181434    c_4370  16:03:23.429229
679529  0_1425  2022-12-20   s_314704     c_304  04:33:37.584749
679530  0_1994  2021-07-16    s_63204    c_2227  20:36:35.350579
679531   1_523  2022-09-28   s_274568    c_3873  01:12:01.973763

[679332 rows x 5 columns]

#### Pour plus de visibilité on sépare l'heure et la date

In [31]:
# vérification de la transaction du produit 731 qui avait un prix négatif
transaction[transaction["id_prod"]==731]

Empty DataFrame
Columns: [id_prod, date, session_id, client_id, heure]
Index: []

In [32]:
# affichage des valeurs doubles
transaction["client_id"].duplicated().sum()

670732

In [33]:
# affichage des valeurs doubles
transaction["id_prod"].duplicated().sum()

676066

In [34]:
# affichage des valeurs doubles
transaction["session_id"].duplicated().sum()

337017

#### Il y a plusieurs valeurs doubles de sessions, de produits et de clients ce qui est logique car des produits identiques sont vendus à différentes personnes, des clients achètent plsuieurs artcicles et des sessions regroupent plusieurs articles 

In [35]:
# description du dataframe
transaction.describe()

id_prod        date session_id client_id            heure
count   679332      679332     679332    679332           679332
unique    3266         730     342315      8600           337816
top      1_369  2021-09-30   s_118668    c_1609  23:37:06.893836
freq      2252        1311         14     25488                6

In [36]:
# verification de la colonne
transaction["id_prod"].nunique()==transaction.shape[0]

False

In [37]:
# verification de la colonne
transaction["session_id"].nunique()==transaction.shape[0]

False

In [38]:
# verification de la colonne
transaction["client_id"].nunique()==transaction.shape[0]

False

# Jointure

In [39]:
# jointure des dataframes product et transaction
lib1=pd.merge(product, 
transaction, how="outer", on=["id_prod"])
lib1

id_prod  price  categ        date session_id client_id            heure
0       0_1421  19.99    0.0  2022-02-20   s_168213    c_6389  17:04:37.204937
1       0_1421  19.99    0.0  2022-11-19   s_299590    c_8364  03:42:20.707442
2       0_1421  19.99    0.0  2021-09-19    s_92304    c_3544  14:58:46.726179
3       0_1421  19.99    0.0  2023-01-11   s_325369    c_1025  16:04:24.367855
4       0_1421  19.99    0.0  2021-08-01    s_70071    c_2298  19:33:18.877168
...        ...    ...    ...         ...        ...       ...              ...
679348  0_2245    NaN    NaN  2021-08-25    s_80395     c_131  09:06:03.504061
679349  0_2245    NaN    NaN  2022-03-06   s_175311    c_4167  19:59:19.462288
679350  0_2245    NaN    NaN  2022-05-16   s_209381    c_4453  11:35:20.319501
679351  0_2245    NaN    NaN  2022-02-11   s_163405    c_1098  09:05:43.952857
679352  0_2245    NaN    NaN  2021-12-14   s_134446    c_4854  22:34:54.589921

[679353 rows x 7 columns]

In [40]:
# vérification des types de colonnes et des valeurs nulles
lib1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 679353 entries, 0 to 679352
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   id_prod     679353 non-null  object 
 1   price       679132 non-null  float64
 2   categ       679132 non-null  float64
 3   date        679332 non-null  object 
 4   session_id  679332 non-null  object 
 5   client_id   679332 non-null  object 
 6   heure       679332 non-null  object 
dtypes: float64(2), object(5)
memory usage: 41.5+ MB


In [41]:
# vérification des valeurs doubles
lib1.duplicated().sum()

0

#### Il n'y a pas de lignes doubles

In [42]:
# jointure des dataframes lib1 et customer
lib2=pd.merge(lib1, 
customer, how="outer", on=["client_id"])
lib2

id_prod  price  categ        date session_id client_id  \
0       0_1421  19.99    0.0  2022-02-20   s_168213    c_6389   
1       0_1421  19.99    0.0  2022-10-20   s_285450    c_6389   
2       0_2131   8.99    0.0  2021-10-09   s_102458    c_6389   
3       0_1635  16.99    0.0  2021-04-28    s_26841    c_6389   
4        0_166   1.83    0.0  2021-07-15    s_62585    c_6389   
...        ...    ...    ...         ...        ...       ...   
679369     NaN    NaN    NaN         NaN        NaN     c_862   
679370     NaN    NaN    NaN         NaN        NaN    c_7584   
679371     NaN    NaN    NaN         NaN        NaN      c_90   
679372     NaN    NaN    NaN         NaN        NaN     c_587   
679373     NaN    NaN    NaN         NaN        NaN    c_3526   

                  heure sex   birth   age  
0       17:04:37.204937   f  1991.0  30.0  
1       17:04:37.204937   f  1991.0  30.0  
2       21:47:06.777914   f  1991.0  30.0  
3       06:08:38.261863   f  1991.0  30.0  
4       10:35:31.526968   f  1991.0  30.0  
...                 ...  ..     ...   ...  
679369              NaN   f  1956.0  65.0  
679370              NaN   f  1960.0  61.0  
679371              NaN   m  2001.0  20.0  
679372              NaN   m  1993.0  28.0  
679373              NaN   m  1956.0  65.0  

[679374 rows x 10 columns]

In [43]:
# changement de type de colonne
lib2["date"] = pd.to_datetime(lib2['date'])

In [44]:
lib2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 679374 entries, 0 to 679373
Data columns (total 10 columns):
 #   Column      Non-Null Count   Dtype         
---  ------      --------------   -----         
 0   id_prod     679353 non-null  object        
 1   price       679132 non-null  float64       
 2   categ       679132 non-null  float64       
 3   date        679332 non-null  datetime64[ns]
 4   session_id  679332 non-null  object        
 5   client_id   679353 non-null  object        
 6   heure       679332 non-null  object        
 7   sex         679353 non-null  object        
 8   birth       679353 non-null  float64       
 9   age         679353 non-null  float64       
dtypes: datetime64[ns](1), float64(4), object(5)
memory usage: 57.0+ MB


#### On voit l'existence de valeurs nulles, nous allons regarder plus en détails.

In [45]:
# affichage des id_prod nuls
nulprod=lib2[lib2["id_prod"].isna()]
print(nulprod.shape)
nulprod

(21, 10)


id_prod  price  categ date session_id client_id heure sex   birth   age
679353     NaN    NaN    NaN  NaT        NaN    c_8253   NaN   f  2001.0  20.0
679354     NaN    NaN    NaN  NaT        NaN    c_3789   NaN   f  1997.0  24.0
679355     NaN    NaN    NaN  NaT        NaN    c_4406   NaN   f  1998.0  23.0
679356     NaN    NaN    NaN  NaT        NaN    c_2706   NaN   f  1967.0  54.0
679357     NaN    NaN    NaN  NaT        NaN    c_3443   NaN   m  1959.0  62.0
679358     NaN    NaN    NaN  NaT        NaN    c_4447   NaN   m  1956.0  65.0
679359     NaN    NaN    NaN  NaT        NaN    c_3017   NaN   f  1992.0  29.0
679360     NaN    NaN    NaN  NaT        NaN    c_4086   NaN   f  1992.0  29.0
679361     NaN    NaN    NaN  NaT        NaN    c_6930   NaN   m  2004.0  17.0
679362     NaN    NaN    NaN  NaT        NaN    c_4358   NaN   m  1999.0  22.0
679363     NaN    NaN    NaN  NaT        NaN    c_8381   NaN   f  1965.0  56.0
679364     NaN    NaN    NaN  NaT        NaN    c_1223   NaN   m  1963.0  58.0
679365     NaN    NaN    NaN  NaT        NaN    c_6862   NaN   f  2002.0  19.0
679366     NaN    NaN    NaN  NaT        NaN    c_5245   NaN   f  2004.0  17.0
679367     NaN    NaN    NaN  NaT        NaN    c_5223   NaN   m  2003.0  18.0
679368     NaN    NaN    NaN  NaT        NaN    c_6735   NaN   m  2004.0  17.0
679369     NaN    NaN    NaN  NaT        NaN     c_862   NaN   f  1956.0  65.0
679370     NaN    NaN    NaN  NaT        NaN    c_7584   NaN   f  1960.0  61.0
679371     NaN    NaN    NaN  NaT        NaN      c_90   NaN   m  2001.0  20.0
679372     NaN    NaN    NaN  NaT        NaN     c_587   NaN   m  1993.0  28.0
679373     NaN    NaN    NaN  NaT        NaN    c_3526   NaN   m  1956.0  65.0

#### On retrouve 21 lignes avec un produit qui n'existe pas

In [46]:
# affichage des client_id nuls 
nulclient=lib2[lib2["client_id"].isna()]
print(nulclient.shape)
nulclient

(21, 10)


id_prod   price  categ date session_id client_id heure  sex  birth  age
607282  0_1016   35.06    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607283  0_1780    1.67    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607284  0_1062   20.08    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607285  0_1119    2.99    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607286  0_1014    1.15    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607287     1_0   31.82    1.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607288  0_1318   20.92    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607289  0_1800   22.05    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607290  0_1645    2.99    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607291   0_322    2.99    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607292  0_1620    0.80    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607293  0_1025   24.99    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607294    2_87  220.99    2.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607295   1_394   39.73    1.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607296    2_72  141.32    2.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607297   0_310    1.94    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607298  0_1624   24.50    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607299    2_86  132.36    2.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607300   0_299   22.99    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607301   0_510   23.66    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607302  0_2308   20.28    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN

#### On retrouve 21 lignes avec un client inexistant

In [47]:
# affichage des clients sans session
clientina= lib2[lib2["client_id"].str.contains("c_", na=False) & lib2["session_id"].isna()]
print(clientina.shape)
clientina.head()

(21, 10)


id_prod  price  categ date session_id client_id heure sex   birth   age
679353     NaN    NaN    NaN  NaT        NaN    c_8253   NaN   f  2001.0  20.0
679354     NaN    NaN    NaN  NaT        NaN    c_3789   NaN   f  1997.0  24.0
679355     NaN    NaN    NaN  NaT        NaN    c_4406   NaN   f  1998.0  23.0
679356     NaN    NaN    NaN  NaT        NaN    c_2706   NaN   f  1967.0  54.0
679357     NaN    NaN    NaN  NaT        NaN    c_3443   NaN   m  1959.0  62.0

#### 21 clients se sont enregistré sur notre site mais n'ont pas encore entamé de session d'achat

### Taux de conversion

In [48]:
# Calcul du taux de conversion
conversion= round((lib2["session_id"].str.contains("s_").count())/(lib2["client_id"].str.contains("c_").count())*100, 3)
conversion

99.997

#### le taux de conversion est pratiquement de 100%, cela signifie que les clients qui s'inscrivent sur le site entament une sessions d'achat.

In [49]:
# affichage des id_prod nuls et session_id nuls
prodnul= lib2[lib2["id_prod"].isna() & lib2["session_id"].isna()]
print(prodnul.shape)
prodnul

(21, 10)


id_prod  price  categ date session_id client_id heure sex   birth   age
679353     NaN    NaN    NaN  NaT        NaN    c_8253   NaN   f  2001.0  20.0
679354     NaN    NaN    NaN  NaT        NaN    c_3789   NaN   f  1997.0  24.0
679355     NaN    NaN    NaN  NaT        NaN    c_4406   NaN   f  1998.0  23.0
679356     NaN    NaN    NaN  NaT        NaN    c_2706   NaN   f  1967.0  54.0
679357     NaN    NaN    NaN  NaT        NaN    c_3443   NaN   m  1959.0  62.0
679358     NaN    NaN    NaN  NaT        NaN    c_4447   NaN   m  1956.0  65.0
679359     NaN    NaN    NaN  NaT        NaN    c_3017   NaN   f  1992.0  29.0
679360     NaN    NaN    NaN  NaT        NaN    c_4086   NaN   f  1992.0  29.0
679361     NaN    NaN    NaN  NaT        NaN    c_6930   NaN   m  2004.0  17.0
679362     NaN    NaN    NaN  NaT        NaN    c_4358   NaN   m  1999.0  22.0
679363     NaN    NaN    NaN  NaT        NaN    c_8381   NaN   f  1965.0  56.0
679364     NaN    NaN    NaN  NaT        NaN    c_1223   NaN   m  1963.0  58.0
679365     NaN    NaN    NaN  NaT        NaN    c_6862   NaN   f  2002.0  19.0
679366     NaN    NaN    NaN  NaT        NaN    c_5245   NaN   f  2004.0  17.0
679367     NaN    NaN    NaN  NaT        NaN    c_5223   NaN   m  2003.0  18.0
679368     NaN    NaN    NaN  NaT        NaN    c_6735   NaN   m  2004.0  17.0
679369     NaN    NaN    NaN  NaT        NaN     c_862   NaN   f  1956.0  65.0
679370     NaN    NaN    NaN  NaT        NaN    c_7584   NaN   f  1960.0  61.0
679371     NaN    NaN    NaN  NaT        NaN      c_90   NaN   m  2001.0  20.0
679372     NaN    NaN    NaN  NaT        NaN     c_587   NaN   m  1993.0  28.0
679373     NaN    NaN    NaN  NaT        NaN    c_3526   NaN   m  1956.0  65.0

In [50]:
# description des clients inactifs
prodnul.describe()

price  categ        birth        age
count    0.0    0.0    21.000000  21.000000
mean     NaN    NaN  1984.380952  36.619048
std      NaN    NaN    19.878320  19.878320
min      NaN    NaN  1956.000000  17.000000
25%      NaN    NaN  1963.000000  20.000000
50%      NaN    NaN  1993.000000  28.000000
75%      NaN    NaN  2001.000000  58.000000
max      NaN    NaN  2004.000000  65.000000

#### Les clients inactifs ont en moyenne 36.6 ans allant de 17 à 65 ans

In [51]:
# affichage des session_id nuls
nulsess=lib2[lib2["session_id"].isna()]
print(nulsess.shape)
nulsess.head()

(42, 10)


id_prod  price  categ date session_id client_id heure  sex  birth  age
607282  0_1016  35.06    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607283  0_1780   1.67    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607284  0_1062  20.08    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607285  0_1119   2.99    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN
607286  0_1014   1.15    0.0  NaT        NaN       NaN   NaN  NaN    NaN  NaN

#### 42 livres n'ont jamais été vendus, il faudra regarder si cela est du à une rupture de stock ou alors un manque d''intérêt de la part du consommateur, il serait alors intéressant de faire une promotion ou une mise en avant.

In [67]:
# affichage du dataframe sans les clients inactifs
librairie=lib2[~lib2["client_id"].str.contains("ct_", na=False)]
print(librairie.shape)
librairie

(679374, 10)


id_prod  price  categ       date session_id client_id            heure  \
0       0_1421  19.99    0.0 2022-02-20   s_168213    c_6389  17:04:37.204937   
1       0_1421  19.99    0.0 2022-10-20   s_285450    c_6389  17:04:37.204937   
2       0_2131   8.99    0.0 2021-10-09   s_102458    c_6389  21:47:06.777914   
3       0_1635  16.99    0.0 2021-04-28    s_26841    c_6389  06:08:38.261863   
4        0_166   1.83    0.0 2021-07-15    s_62585    c_6389  10:35:31.526968   
...        ...    ...    ...        ...        ...       ...              ...   
679369     NaN    NaN    NaN        NaT        NaN     c_862              NaN   
679370     NaN    NaN    NaN        NaT        NaN    c_7584              NaN   
679371     NaN    NaN    NaN        NaT        NaN      c_90              NaN   
679372     NaN    NaN    NaN        NaT        NaN     c_587              NaN   
679373     NaN    NaN    NaN        NaT        NaN    c_3526              NaN   

       sex   birth   age  
0        f  1991.0  30.0  
1        f  1991.0  30.0  
2        f  1991.0  30.0  
3        f  1991.0  30.0  
4        f  1991.0  30.0  
...     ..     ...   ...  
679369   f  1956.0  65.0  
679370   f  1960.0  61.0  
679371   m  2001.0  20.0  
679372   m  1993.0  28.0  
679373   m  1956.0  65.0  

[679374 rows x 10 columns]

#### les clients inactifs ont été enlevé de l'analyse 

In [68]:
# vérification des types de colonnes et des valeurs nulles
librairie.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 679374 entries, 0 to 679373
Data columns (total 10 columns):
 #   Column      Non-Null Count   Dtype         
---  ------      --------------   -----         
 0   id_prod     679353 non-null  object        
 1   price       679132 non-null  float64       
 2   categ       679132 non-null  float64       
 3   date        679332 non-null  datetime64[ns]
 4   session_id  679332 non-null  object        
 5   client_id   679353 non-null  object        
 6   heure       679332 non-null  object        
 7   sex         679353 non-null  object        
 8   birth       679353 non-null  float64       
 9   age         679353 non-null  float64       
dtypes: datetime64[ns](1), float64(4), object(5)
memory usage: 57.0+ MB


In [54]:
# vérification des valeurs doubles
librairie.duplicated().sum()

0

In [69]:
# dataframe sans les valeurs nulles
librairie=librairie[librairie["id_prod"].notna() & librairie["client_id"].notna() & librairie["session_id"].notna()]
librairie

id_prod  price  categ       date session_id client_id            heure  \
0       0_1421  19.99    0.0 2022-02-20   s_168213    c_6389  17:04:37.204937   
1       0_1421  19.99    0.0 2022-10-20   s_285450    c_6389  17:04:37.204937   
2       0_2131   8.99    0.0 2021-10-09   s_102458    c_6389  21:47:06.777914   
3       0_1635  16.99    0.0 2021-04-28    s_26841    c_6389  06:08:38.261863   
4        0_166   1.83    0.0 2021-07-15    s_62585    c_6389  10:35:31.526968   
...        ...    ...    ...        ...        ...       ...              ...   
679348   2_163  68.99    2.0 2022-01-28   s_156517    c_7739  16:51:07.389515   
679349   2_101  63.99    2.0 2021-07-21    s_65192    c_7089  09:11:12.619736   
679350   2_101  63.99    2.0 2022-12-21   s_315267    c_7089  09:11:12.619736   
679351   2_101  63.99    2.0 2022-10-21   s_285788    c_7089  09:11:12.619736   
679352   2_101  63.99    2.0 2022-03-21   s_182240    c_7089  09:11:12.619736   

       sex   birth   age  
0        f  1991.0  30.0  
1        f  1991.0  30.0  
2        f  1991.0  30.0  
3        f  1991.0  30.0  
4        f  1991.0  30.0  
...     ..     ...   ...  
679348   m  1997.0  24.0  
679349   m  2002.0  19.0  
679350   m  2002.0  19.0  
679351   m  2002.0  19.0  
679352   m  2002.0  19.0  

[679332 rows x 10 columns]

#### Les valeurs nulles ont été enlevées afin d'avoir une analyse plus pertinente

In [70]:
# création des colonnes année, mois, jour et nom du jour
librairie["année"]=librairie["date"].dt.year
librairie["mois"]=librairie["date"].dt.month
librairie["jour"]=librairie["date"].dt.day
librairie["nom_jour"]=librairie["date"].dt.day_name()
librairie

<ipython-input-70-176f515e8395>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  librairie["année"]=librairie["date"].dt.year
<ipython-input-70-176f515e8395>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  librairie["mois"]=librairie["date"].dt.month
<ipython-input-70-176f515e8395>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/

id_prod  price  categ       date session_id client_id            heure  \
0       0_1421  19.99    0.0 2022-02-20   s_168213    c_6389  17:04:37.204937   
1       0_1421  19.99    0.0 2022-10-20   s_285450    c_6389  17:04:37.204937   
2       0_2131   8.99    0.0 2021-10-09   s_102458    c_6389  21:47:06.777914   
3       0_1635  16.99    0.0 2021-04-28    s_26841    c_6389  06:08:38.261863   
4        0_166   1.83    0.0 2021-07-15    s_62585    c_6389  10:35:31.526968   
...        ...    ...    ...        ...        ...       ...              ...   
679348   2_163  68.99    2.0 2022-01-28   s_156517    c_7739  16:51:07.389515   
679349   2_101  63.99    2.0 2021-07-21    s_65192    c_7089  09:11:12.619736   
679350   2_101  63.99    2.0 2022-12-21   s_315267    c_7089  09:11:12.619736   
679351   2_101  63.99    2.0 2022-10-21   s_285788    c_7089  09:11:12.619736   
679352   2_101  63.99    2.0 2022-03-21   s_182240    c_7089  09:11:12.619736   

       sex   birth   age  année  mois  jour   nom_jour  
0        f  1991.0  30.0   2022     2    20     Sunday  
1        f  1991.0  30.0   2022    10    20   Thursday  
2        f  1991.0  30.0   2021    10     9   Saturday  
3        f  1991.0  30.0   2021     4    28  Wednesday  
4        f  1991.0  30.0   2021     7    15   Thursday  
...     ..     ...   ...    ...   ...   ...        ...  
679348   m  1997.0  24.0   2022     1    28     Friday  
679349   m  2002.0  19.0   2021     7    21  Wednesday  
679350   m  2002.0  19.0   2022    12    21  Wednesday  
679351   m  2002.0  19.0   2022    10    21     Friday  
679352   m  2002.0  19.0   2022     3    21     Monday  

[679332 rows x 14 columns]

In [71]:
librairie.columns

Index(['id_prod', 'price', 'categ', 'date', 'session_id', 'client_id', 'heure',
       'sex', 'birth', 'age', 'année', 'mois', 'jour', 'nom_jour'],
      dtype='object')

In [72]:
# mise en ordre des colones
librairie=librairie[['id_prod', 'price', 'categ', 'session_id', 'date', 'heure', 'nom_jour', 'jour', 'mois', 'année',  'client_id', 'sex', 'birth', 'age']]
librairie

id_prod  price  categ session_id       date            heure  \
0       0_1421  19.99    0.0   s_168213 2022-02-20  17:04:37.204937   
1       0_1421  19.99    0.0   s_285450 2022-10-20  17:04:37.204937   
2       0_2131   8.99    0.0   s_102458 2021-10-09  21:47:06.777914   
3       0_1635  16.99    0.0    s_26841 2021-04-28  06:08:38.261863   
4        0_166   1.83    0.0    s_62585 2021-07-15  10:35:31.526968   
...        ...    ...    ...        ...        ...              ...   
679348   2_163  68.99    2.0   s_156517 2022-01-28  16:51:07.389515   
679349   2_101  63.99    2.0    s_65192 2021-07-21  09:11:12.619736   
679350   2_101  63.99    2.0   s_315267 2022-12-21  09:11:12.619736   
679351   2_101  63.99    2.0   s_285788 2022-10-21  09:11:12.619736   
679352   2_101  63.99    2.0   s_182240 2022-03-21  09:11:12.619736   

         nom_jour  jour  mois  année client_id sex   birth   age  
0          Sunday    20     2   2022    c_6389   f  1991.0  30.0  
1        Thursday    20    10   2022    c_6389   f  1991.0  30.0  
2        Saturday     9    10   2021    c_6389   f  1991.0  30.0  
3       Wednesday    28     4   2021    c_6389   f  1991.0  30.0  
4        Thursday    15     7   2021    c_6389   f  1991.0  30.0  
...           ...   ...   ...    ...       ...  ..     ...   ...  
679348     Friday    28     1   2022    c_7739   m  1997.0  24.0  
679349  Wednesday    21     7   2021    c_7089   m  2002.0  19.0  
679350  Wednesday    21    12   2022    c_7089   m  2002.0  19.0  
679351     Friday    21    10   2022    c_7089   m  2002.0  19.0  
679352     Monday    21     3   2022    c_7089   m  2002.0  19.0  

[679332 rows x 14 columns]

In [73]:
# vérification des jours
librairie["jour"].unique()

array([20,  9, 28, 15, 10, 29, 30,  1, 22,  4,  6, 13, 11, 21,  3, 19, 17,
       18,  2,  8,  5, 26, 16, 27, 25, 24,  7, 31, 12, 14, 23],
      dtype=int64)

In [74]:
# vérification des mois
librairie["mois"].unique()

array([ 2, 10,  4,  7,  6,  1, 11, 12,  5,  8,  3,  9], dtype=int64)

In [75]:
# vérification des années
librairie["année"].unique()

array([2022, 2021, 2023], dtype=int64)

In [76]:
# sauvegarde du fichier de nettoyage
librairie.to_csv("librairie_nettoyage.csv", index=False)